### PR TITLE
Add end locations to all nodes

### PR DESCRIPTION
**Pull request by charliermarsh**
# Summary

I'm finding myself needing end locations for AST nodes (similar to in the Python AST). Is there interest in enabling this? If so, happy to clean up and test further.

### CODE REVIEW

**Comment by youknowone** [comment]
Please feel free to let me know if you need any help

**Comment by charliermarsh** [comment]
@youknowone - Just rebased. I've been using this code successfully in Ruff for a while now, would love to get it merged if you're open to it! Let me know what needs fixing / addressing :)

**Comment by youknowone** [comment]
@charliermarsh I didn't check the details yet, but could you check the failing tests?

**Comment by charliermarsh** [comment]
@youknowone - Absolutely, will do, thanks.

**Comment by charliermarsh** [comment]
@youknowone - Tests passing!

**Comment by youknowone** [comment]
Because we are trying to leave our own footprints to CPython codes as less as possible, I reverted the last commit but tried to change end_location to be optional.

**Comment by charliermarsh** [comment]
I can change it to `Optional` (but populated everywhere, as it is now) if you'd prefer?

**Comment by charliermarsh** [comment]
Or are you suggesting something else? Using `Optional` is less convenient for me as a "client" of the parser but I know the spec marks it as optional :)

**Comment by charliermarsh** [comment]
I can try to make that change today.

**Comment by charliermarsh** [comment]
In practice, does CPython ever omit `end_lineno` and `end_col_offset` when you go through `ast.parse`? Or are they just optional when creating AST nodes "manually"?

**Comment by youknowone** [comment]
@charliermarsh because I made incomplete commit last push, you can work on the new one as a base.

I am sorry, I don't know well about CPython ast that much.
cc @fanninpm @coolreader18

**Comment by charliermarsh** [comment]
@youknowone - Hoping I didn't override your work. Here's the first part.

File: Lib/test/test_ast.py [hunk 1]
```diff
@@ -2307,8 +2307,6 @@ class C:
         cdef = ast.parse(s).body[0]
         self.assertEqual(ast.get_source_segment(s, cdef.body[0], padded=True), s_method)
 
-    # TODO: RUSTPYTHON
-    @unittest.expectedFailure
```
Comment (charliermarsh) [inline Lib/test/test_ast.py:2311]: @youknowone - All other test changes were now able to be reverted.

File: compiler/ast/src/ast_gen.rs [hunk 2]
```diff
@@ -10,14 +10,16 @@ type Ident = String;
...
 pub struct Located<T, U = ()> {
     pub location: Location,
+    pub end_location: Option<Location>,
     pub custom: U,
     pub node: T,
 }
 
 impl<T> Located<T> {
-    pub fn new(location: Location, node: T) -> Self {
+    pub fn new(location: Location, end_location: Location, node: T) -> Self {
```
Comment (charliermarsh) [inline compiler/ast/src/ast_gen.rs:19]: I kept `Location` as a required field in `::new` for now. (We can omit it in special cases by going through the Struct directly, as in `ast.rs`.)

Comment (charliermarsh) [reply compiler/ast/src/ast_gen.rs:19]: I could see arguments either way here. On the one hand, it's odd that the field you pass in differs from what's stored on the struct. On the other hand, this encourages good behavior _and_ reduces the number of code changes required.

Comment (youknowone) [reply compiler/ast/src/ast_gen.rs:19]: I agree this approach and the idea encouraging good behavior.

File: compiler/parser/src/context.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
+use rustpython_ast::{Expr, ExprContext, ExprKind};
+
+pub fn set_context(expr: Expr, ctx: ExprContext) -> Expr {
```
Comment (youknowone) [inline compiler/parser/src/context.rs:3]: is this necessary part of end_location?

**Review by youknowone — APPROVED** [review]

File: compiler/parser/src/context.rs (hunk 3 above)
Comment (charliermarsh) [reply compiler/parser/src/context.rs:3]: Oh no, sorry, I need to revert this. Not sure how it got included.

Comment (charliermarsh) [reply compiler/parser/src/context.rs]: (Removed.)

**Comment by youknowone** [comment]
Awesome! Thank you!

**Pull request merged**